x/fileurl.go

```diff
@@ -0,0 +1,110 @@
+package x
+
+import (
+	"net/url"
+	"path/filepath"
+	"regexp"
+	"runtime"
+	"strings"
+
+	"github.com/ory/x/logrusx"
+)
+
+// winPathRegex is a regex for [DRIVE-LETTER]:
+var winPathRegex = regexp.MustCompile("^[A-Za-z]:.*")
+
+// GetURLFilePath returns the path of a URL that is compatible with the runtime os filesystem
+func GetURLFilePath(u *url.URL) string {
+	if u == nil {
+		return ""
+	}
+
+	if u.Scheme != "file" && u.Scheme != "" {
+		return u.Path
+	}
+	fPath := u.Path
+	if runtime.GOOS == "windows" {
+		if u.Host != "" {
+			// Make UNC Path
+			s := string(filepath.Separator)
+			fPath = s + s + u.Host + filepath.FromSlash(fPath)
+			return fPath
+		}
+		if winPathRegex.MatchString(fPath[1:]) {
+			// On Windows we should remove the initial path separator in case this
+			// is a normal path (for example: "\c:\" -> "c:\"")
+			fPath = stripFistPathSeparators(fPath)
+		}
+	}
+	return filepath.FromSlash(fPath)
+}
+
+// ParseURL parses rawurl into a URL structure with special handling for file:// URLs
+func ParseURL(rawurl string) (*url.URL, error) {
+	lcRawurl := strings.ToLower(rawurl)
+	if strings.Index(lcRawurl, "file:///") == 0 {
+		return url.Parse("file:///" + toSlash(rawurl[8:]))
+	}
+	if strings.Index(lcRawurl, "file://") == 0 {
+		// Normally the first part after file:// is a hostname, but since
+		// this is often misused we interpret the URL like a normal path
+		// by removing the "file://" from the beginning
+		rawurl = rawurl[7:]
+	}
+	if winPathRegex.MatchString(rawurl) {
+		// Windows path
+		return url.Parse("file:///" + toSlash(rawurl))
+	}
+	if strings.Index(lcRawurl, "\\\\") == 0 {
+		// Windows UNC path
+		// We extract the hostname and creates an appropriate file:// URL
+		// based on the hostname and the path
+		parts := strings.Split(filepath.FromSlash(rawurl), "\\")
```

Comment:
On POSIX, `filepath.FromSlash` will result in a path with slashes again and the split will match literal backslashes (not that any sane person would use backslashes in directory/file names but it should not match those...)
I would say also the condition at L58 should probably not be true on POSIX. I feel like this might open an attack vector 10 years in the future where somebody uses `\` in filenames to circumvent some check.